### DISCUSSION TITLE
NextJS + Zustand localStorage persist middleware causing hydration errors

### DISCUSSION THREAD

**Discussion by chrber04** [pmndrs/zustand#1382] [Q&A]
I am unable to persist a few states like the theme, layout etc in my Nextjs app using Zustand with the persist middleware. I'm using localStorage as the storage option. After refreshing the window, it causes a content mismatch error:

`Error: Hydration failed because the initial UI does not match what was rendered on the server.`

If this already is a known issue, how would I go about using the persist middleware whilst avoiding this? 

Thanks

**Comment by dai-shi** [comment]
This is an expected behavior because what server knows and client knows is different (=mismatch). You could double render on client or don't let server render that depends on client data.

I'm not exactly sure how zustand users work around this. Let's see if someone responds.

**Comment by charkour** [comment]
With zustand v3, we use a custom storage engine and make get localStorage an async method. 

Haven't tried this workaround yet with v4, but it works with v3, and the latest next and react versions. 

See this comment: https://github.com/pmndrs/zustand/issues/1145#issuecomment-1202871214
